src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp

```diff
@@ -35,18 +36,25 @@
 #include "utilities/ticks.hpp"
 
 class G1ResetPinnedClosure : public HeapRegionClosure {
-  G1CMBitMap* _bitmap;
+  G1FullCollector* _collector;
 
 public:
-  G1ResetPinnedClosure(G1CMBitMap* bitmap) : _bitmap(bitmap) { }
+  G1ResetPinnedClosure(G1FullCollector* collector) : _collector(collector) { }
 
   bool do_heap_region(HeapRegion* r) {
-    if (!r->is_pinned()) {
+    uint hr_index = r->hrm_index();
+    // In the prepare phase, we "pin" the regions with high survival ratio
+    // by _region_attr_table, so here we use _region_attr_table rather than
+    // HeapRegion itself to tell whether a region is pinned.
+    if (!_collector->is_in_pinned_or_closed(hr_index)) {
```

Comment:
The original code uses only `r->is_pinned()`, not `r->is_pinned_or_closed()`.
This predicate is wrong, `closed` regions may not be pinned (but can be, and in most cases are).

Reply:
I'm not sure, but per code suggests in src/hotspot/share/gc/g1/heapRegionType.hpp, seems closed region should be a pinned as well. Or miss something? Would you mind to clarify this point?
    ArchiveMask           = 32,
    OpenArchiveTag        = ArchiveMask | PinnedMask,
    ClosedArchiveTag      = ArchiveMask | PinnedMask + 1

Reply:
Just because a `ClosedArchive` region is by default pinned at this time, it does not mean it will be in the future. There are no immediate plans for that, but I can think of some edge cases where it could be useful.

The correct (and intended) predicate for this check in this context is `is_pinned()`, although at the moment they have the same effect.

Reply:
I see, thanks for clarifying.

Reply:
For "is_pinned()", do you mean the method in HeapRegion?

Reply:
Attribute table - full gc is only concerned with the attribute table.

Reply:
I see, Thanks